go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,270 @@
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backmode mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		_ = hacluster.DemoteOVNDbs(invalidIPAddres, "6641", "6642")
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterNewLeader := func(identity string) {
+		logrus.Infof("The new leader is : " + identity)
+		if hacluster.nodeName == identity {
+			_ = hacluster.ConfigureAsActive(identity)
+		} else {
+			_ = hacluster.ConfigureAsStandby(identity)
+		}
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDeadline,
+		RetryPeriod:   retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {},
+			OnStoppedLeading: func() {},
+			OnNewLeader:      HAClusterNewLeader,
+		},
+	}
+
+	go leaderelection.RunOrDie(context.Background(), lec)
+
+	return nil
+}
+
+// ConfigureAsActive configures the node as active.
+func (hacluster *OvnHAMasterController) ConfigureAsActive(masterNodeName string) error {
+	if hacluster.manageDBServers {
+		// Find the endpoint for the service
+		ep, err := hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+		createEp := true
+		if err == nil {
+			if ep.Name == ovnkubeDbEp {
+				createEp = false
+			}
+		}
+
+		const (
+			nbPort int32 = 6641
+			sbPort int32 = 6642
+		)
+
+		ovndbEp := v1.Endpoints{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: config.Kubernetes.OVNConfigNamespace,
+				Name:      ovnkubeDbEp,
+			},
+			Subsets: []v1.EndpointSubset{
+				{
+					Addresses: []v1.EndpointAddress{
+						{IP: config.Kubernetes.OVNKubePodIP},
+					},
+					Ports: []v1.EndpointPort{
+						{
+							Name: "north",
+							Port: nbPort,
+						},
+						{
+							Name: "south",
+							Port: sbPort,
+						},
+					},
+				},
+			},
+		}
+
+		if createEp {
+			_, err := hacluster.cluster.Kube.CreateEndpoint(config.Kubernetes.OVNConfigNamespace, &ovndbEp)
+			if err != nil {
+				//TODO: Handle this situation properly.
+				logrus.Infof("%s Endpount Create failed", ovnkubeDbEp)
+			}
+		} else {
+			_, err := hacluster.cluster.Kube.UpdateEndpoint(config.Kubernetes.OVNConfigNamespace, &ovndbEp)
+			if err != nil {
+				//TODO: Handle this situation properly.
+				logrus.Infof("%s Endpount Update failed", ovnkubeDbEp)
```

Comment:
Just Exit. We'll get restarted by kube and enter leader election again.

Reply:
Ack.